Use OpenSSL ASN1_STRING accessors instead of reaching into the
ASN1_TIME struct. OpenSSL 4.0 made ASN1_STRING opaque, so
timeobj->data / timeobj->type no longer compile.

Backport of https://github.com/Cisco-Talos/clamav/pull/1642
(1.6 later removed this private helper entirely).

--- clamav-1.5.4/libclamav/crypto.c
+++ clamav-1.5.4/libclamav/crypto.c
@@ -52,6 +52,7 @@
 #include <fcntl.h>
 
 #include <openssl/evp.h>
+#include <openssl/asn1.h>
 
 #if OPENSSL_VERSION_NUMBER < 0x10100000L
 #define X509_CRL_get0_nextUpdate X509_CRL_get_nextUpdate
@@ -1756,9 +1757,10 @@
 
 struct tm *cl_ASN1_GetTimeT(ASN1_TIME *timeobj)
 {
-    struct tm *t;
-    char *str;
-    const char *fmt = NULL;
+    struct tm *ret = NULL;
+    struct tm *t   = NULL;
+    char *str      = NULL;
+    const char *data, *fmt = NULL;
     time_t localt;
 #ifdef _WIN32
     struct tm localtm, *ltm;
@@ -1766,18 +1768,21 @@
     struct tm localtm;
 #endif
 
-    if (!(timeobj) || !(timeobj->data))
-        return NULL;
+    if (timeobj == NULL || (data = (const char *)ASN1_STRING_get0_data(timeobj)) == NULL)
+        goto done;
 
-    str = (char *)(timeobj->data);
+    str = CLI_STRNDUP(data, ASN1_STRING_length(timeobj));
+    if (NULL == str)
+        goto done;
+
     if (strlen(str) < 12)
-        return NULL;
+        goto done;
 
     t = (struct tm *)calloc(1, sizeof(struct tm));
     if (!(t))
-        return NULL;
+        goto done;
 
-    if (timeobj->type == V_ASN1_UTCTIME) {
+    if (ASN1_STRING_type(timeobj) == V_ASN1_UTCTIME) {
         /* two digit year */
         fmt = "%y%m%d%H%M%S";
         if (str[3] == '0') {
@@ -1786,7 +1791,7 @@
         } else {
             str[3]--;
         }
-    } else if (timeobj->type == V_ASN1_GENERALIZEDTIME) {
+    } else if (ASN1_STRING_type(timeobj) == V_ASN1_GENERALIZEDTIME) {
         /* four digit year */
         fmt = "%Y%m%d%H%M%S";
         if (str[5] == '0') {
@@ -1797,15 +1802,11 @@
         }
     }
 
-    if (!(fmt)) {
-        free(t);
-        return NULL;
-    }
+    if (!(fmt))
+        goto done;
 
-    if (!strptime(str, fmt, t)) {
-        free(t);
-        return NULL;
-    }
+    if (!strptime(str, fmt, t))
+        goto done;
 
     /* Convert to local time */
     localt = time(NULL);
@@ -1816,7 +1817,15 @@
     localtime_r(&localt, &localtm);
 #endif
     t->tm_isdst = localtm.tm_isdst;
-    return t;
+
+    ret = t;
+    t   = NULL;
+
+done:
+    free(t);
+    free(str);
+
+    return ret;
 }
 
 X509_CRL *cl_load_crl(const char *file)
